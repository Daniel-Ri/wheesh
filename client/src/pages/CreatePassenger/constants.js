export const CREATE_PASSENGER = 'CreatePassenger/CREATE_PASSEGER';
