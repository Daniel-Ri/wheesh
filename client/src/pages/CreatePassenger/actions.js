import { CREATE_PASSENGER } from './constants';

export const createPassenger = (inputs, handleSuccess, handleError) => ({
  type: CREATE_PASSENGER,
  inputs,
  handleSuccess,
  handleError,
});
