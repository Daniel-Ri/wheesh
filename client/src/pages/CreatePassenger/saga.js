import { takeLatest, call } from 'redux-saga/effects';
import { createPassenger } from '@domain/api';
import { CREATE_PASSENGER } from './constants';

function* doCreatePassenger({ inputs, handleSuccess, handleError }) {
  try {
    yield call(createPassenger, inputs);
    yield call(handleSuccess);
  } catch (error) {
    // error
    yield call(handleError, error.response.data.message);
  }
}

export default function* createPassengerSaga() {
  yield takeLatest(CREATE_PASSENGER, doCreatePassenger);
}
