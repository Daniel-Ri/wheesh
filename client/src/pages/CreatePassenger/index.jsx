/* eslint-disable jsx-a11y/label-has-associated-control */
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { FormattedMessage, injectIntl } from 'react-intl';

import { useNavigate } from 'react-router-dom';
import BackBtn from '@components/BackBtn';
import { useState } from 'react';
import toast from 'react-hot-toast';
import DatePicker from 'react-datepicker';
import 'react-datepicker/dist/react-datepicker.css';
import './custom.css';

import { Button } from '@mui/material';
import { createPassenger } from './actions';

import classes from './style.module.scss';

const CreatePassenger = ({ intl: { formatMessage } }) => {
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const [inputs, setInputs] = useState({
    gender: '',
    dateOfBirth: '',
    idCard: '',
    name: '',
    email: '',
  });

  const handleInputChange = (e) => {
    setInputs((prev) => ({ ...prev, [e.target.name]: e.target.value }));
  };

  const validateIdCard = () => {
    if (!inputs.idCard) {
      toast.error(formatMessage({ id: 'app_must_insert_id_card' }));
      return false;
    }
    if (inputs.idCard.length !== 16 || !/^\d+$/.test(inputs.idCard)) {
      toast.error(formatMessage({ id: 'app_incorrect_id_card_format' }));
      return false;
    }

    return true;
  };

  const validateEmail = () => {
    if (!inputs.email) {
      return true;
    }

    if (!/^[a-zA-Z0-9.!#$%&â€™*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/.test(inputs.email)) {
      toast.error(formatMessage({ id: 'app_incorrect_email_format' }));
      return false;
    }

    return true;
  };

  const handleFocusOut = (e) => {
    if (e.target.name === 'idCard') {
      validateIdCard();
    }

    if (e.target.name === 'email') {
      validateEmail();
    }
  };

  const validateInputs = () => {
    if (!validateIdCard()) return false;
    if (!validateEmail()) return false;
    return true;
  };

  const handleSuccess = () => {
    toast.success(formatMessage({ id: 'app_success_add_passenger' }));
    navigate('/myPassengers');
  };

  const handleError = (errorMsg) => {
    toast.error(errorMsg);
  };

  const handleSubmit = () => {
    if (!validateInputs()) return;

    const formattedInputs = { ...inputs };
    formattedInputs.dateOfBirth = new Date(inputs.dateOfBirth).toDateString();

    if (!inputs.email) delete formattedInputs.email;

    dispatch(createPassenger(formattedInputs, handleSuccess, handleError));
  };

  return (
    <main className={classes.main}>
      <div className={classes.container}>
        <header>
          <BackBtn handleClickBack={() => navigate('/myPassengers')} />
          <h1>
            <FormattedMessage id="app_add_passenger" />
          </h1>
        </header>

        <form>
          <div className={classes.header}>
            <FormattedMessage id="app_personal_information" />
          </div>
          <div className={classes.input}>
            <label htmlFor="gender">
              <FormattedMessage id="app_gender" />
            </label>
            <div className={classes.radios}>
              <div className={classes.radio} onClick={() => setInputs((prev) => ({ ...prev, gender: 'Male' }))}>
                <input
                  type="radio"
                  name="gender"
                  value="Male"
                  checked={inputs.gender === 'Male'}
                  onChange={handleInputChange}
                />
                <label>
                  <FormattedMessage id="app_male" />
                </label>
              </div>
              <div className={classes.radio} onClick={() => setInputs((prev) => ({ ...prev, gender: 'Female' }))}>
                <input
                  type="radio"
                  name="gender"
                  value="Female"
                  checked={inputs.gender === 'Female'}
                  onChange={handleInputChange}
                />
                <label>
                  <FormattedMessage id="app_female" />
                </label>
              </div>
            </div>
          </div>
          <div className={classes.input}>
            <label htmlFor="dateOfBirth">
              <FormattedMessage id="app_date_of_birth" />
            </label>
            <DatePicker
              name="dateOfBirth"
              placeholderText={formatMessage({ id: 'app_select_the_date_of_birth' })}
              selected={inputs.dateOfBirth}
              showMonthDropdown
              dropdownMode="select"
              showYearDropdown
              dateFormat="dd/MM/yyyy"
              className={classes.datePicker}
              onChange={(date) => setInputs((prev) => ({ ...prev, dateOfBirth: date }))}
            />
          </div>

          <div className={classes.header}>
            <FormattedMessage id="app_certificate_information" />
          </div>
          <div className={classes.input}>
            <label htmlFor="idCard">
              <FormattedMessage id="app_id_card" />
            </label>
            <input
              type="text"
              name="idCard"
              id="idCard"
              value={inputs.idCard}
              onChange={handleInputChange}
              onBlur={handleFocusOut}
              placeholder={formatMessage({ id: 'app_please_enter_indonesia_id' })}
              autoComplete="off"
            />
          </div>
          <div className={classes.input}>
            <label htmlFor="name">
              <FormattedMessage id="app_name" />
            </label>
            <input
              type="text"
              name="name"
              id="name"
              value={inputs.name}
              onChange={handleInputChange}
              placeholder={formatMessage({ id: 'app_enter_name_on_id_card' })}
              autoComplete="off"
            />
          </div>

          <div className={classes.header}>
            <FormattedMessage id="app_contact_information" />
          </div>
          <div className={classes.input}>
            <label htmlFor="email">E-mail</label>
            <input
              type="text"
              name="email"
              id="email"
              value={inputs.email}
              onChange={handleInputChange}
              onBlur={handleFocusOut}
              placeholder={formatMessage({ id: 'app_optional_enter_email_address' })}
              autoComplete="off"
            />
          </div>
        </form>

        <div className={classes.buttons}>
          <Button variant="contained" className={classes.submit} onClick={handleSubmit}>
            <FormattedMessage id="app_submit" />
          </Button>
        </div>
      </div>
    </main>
  );
};

CreatePassenger.propTypes = {
  intl: PropTypes.object,
};

export default injectIntl(CreatePassenger);
