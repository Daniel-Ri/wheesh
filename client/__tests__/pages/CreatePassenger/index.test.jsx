import { render, fireEvent } from '@utils/testHelper';
import CreatePassenger from '@pages/CreatePassenger';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  useNavigate: () => mockNavigate,
}));

let wrapper;

beforeEach(() => {
  wrapper = render(<CreatePassenger />);
});

describe('Page CreatePassenger', () => {
  test('Renders CreatePassenger page', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('CreatePassenger')).toBeInTheDocument();
  });

  test('Should call navigate when BackBtn clicked', () => {
    const { getByTestId } = wrapper;
    const backButton = getByTestId('BackBtn');
    fireEvent.click(backButton);
    expect(mockNavigate).toHaveBeenCalledWith('/myPassengers');
  });

  test('Should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
