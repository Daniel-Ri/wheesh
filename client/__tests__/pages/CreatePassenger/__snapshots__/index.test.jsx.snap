// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page CreatePassenger Should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="main"
        data-testid="CreatePassenger"
      >
        <div
          class="container"
        >
          <header>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium btn css-1e6y48t-MuiButtonBase-root-MuiButton-root"
              data-testid="BackBtn"
              tabindex="0"
              type="button"
            >
              Kembali
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <h1>
              Tambah Penumpang
            </h1>
          </header>
          <form>
            <div
              class="header"
            >
              Informasi Pribadi
            </div>
            <div
              class="input"
            >
              <label
                for="gender"
              >
                Jenis Kelamin
              </label>
              <div
                class="radios"
              >
                <div
                  class="radio"
                >
                  <input
                    name="gender"
                    type="radio"
                    value="Male"
                  />
                  <label>
                    Pria
                  </label>
                </div>
                <div
                  class="radio"
                >
                  <input
                    name="gender"
                    type="radio"
                    value="Female"
                  />
                  <label>
                    Wanita
                  </label>
                </div>
              </div>
            </div>
            <div
              class="input"
            >
              <label
                for="dateOfBirth"
              >
                Tanggal Kelahiran
              </label>
              <div
                class="datePickerWrapper"
              >
                <div
                  class="react-datepicker-wrapper"
                >
                  <div
                    class="react-datepicker__input-container"
                  >
                    <input
                      class="datePicker"
                      name="dateOfBirth"
                      placeholder="Pilih tanggal kelahiran"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="header"
            >
              Informasi Sertifikat
            </div>
            <div
              class="input"
            >
              <label
                for="idCard"
              >
                Kartu Identitas
              </label>
              <input
                autocomplete="off"
                id="idCard"
                name="idCard"
                placeholder="Silakan masukkan KTP Indonesia"
                type="text"
                value=""
              />
            </div>
            <div
              class="input"
            >
              <label
                for="name"
              >
                Nama
              </label>
              <input
                autocomplete="off"
                id="name"
                name="name"
                placeholder="Masukkan nama sesuai KTP"
                type="text"
                value=""
              />
            </div>
            <div
              class="header"
            >
              Informasi Kontak
            </div>
            <div
              class="input"
            >
              <label
                for="email"
              >
                E-mail
              </label>
              <input
                autocomplete="off"
                id="email"
                name="email"
                placeholder="(Opsional) Masukkan alamat email"
                type="text"
                value=""
              />
            </div>
          </form>
          <div
            class="buttons"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium submit css-sghohy-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Kirim
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="main"
      data-testid="CreatePassenger"
    >
      <div
        class="container"
      >
        <header>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium btn css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            data-testid="BackBtn"
            tabindex="0"
            type="button"
          >
            Kembali
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <h1>
            Tambah Penumpang
          </h1>
        </header>
        <form>
          <div
            class="header"
          >
            Informasi Pribadi
          </div>
          <div
            class="input"
          >
            <label
              for="gender"
            >
              Jenis Kelamin
            </label>
            <div
              class="radios"
            >
              <div
                class="radio"
              >
                <input
                  name="gender"
                  type="radio"
                  value="Male"
                />
                <label>
                  Pria
                </label>
              </div>
              <div
                class="radio"
              >
                <input
                  name="gender"
                  type="radio"
                  value="Female"
                />
                <label>
                  Wanita
                </label>
              </div>
            </div>
          </div>
          <div
            class="input"
          >
            <label
              for="dateOfBirth"
            >
              Tanggal Kelahiran
            </label>
            <div
              class="datePickerWrapper"
            >
              <div
                class="react-datepicker-wrapper"
              >
                <div
                  class="react-datepicker__input-container"
                >
                  <input
                    class="datePicker"
                    name="dateOfBirth"
                    placeholder="Pilih tanggal kelahiran"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="header"
          >
            Informasi Sertifikat
          </div>
          <div
            class="input"
          >
            <label
              for="idCard"
            >
              Kartu Identitas
            </label>
            <input
              autocomplete="off"
              id="idCard"
              name="idCard"
              placeholder="Silakan masukkan KTP Indonesia"
              type="text"
              value=""
            />
          </div>
          <div
            class="input"
          >
            <label
              for="name"
            >
              Nama
            </label>
            <input
              autocomplete="off"
              id="name"
              name="name"
              placeholder="Masukkan nama sesuai KTP"
              type="text"
              value=""
            />
          </div>
          <div
            class="header"
          >
            Informasi Kontak
          </div>
          <div
            class="input"
          >
            <label
              for="email"
            >
              E-mail
            </label>
            <input
              autocomplete="off"
              id="email"
              name="email"
              placeholder="(Opsional) Masukkan alamat email"
              type="text"
              value=""
            />
          </div>
        </form>
        <div
          class="buttons"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium submit css-sghohy-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Kirim
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
